File: lintcode/two-pointer/3Sum.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	res := [][]int{}
	i, j := 0, len(nums)-1
	for ; i < j; {
		tmp := [][]int{}
		t := nums[i]
		twoSum(nums[i+1:], &tmp, -nums[i])
		for k := 0; k < len(tmp); k++ {
			res = append(res, append(tmp[k], nums[i]))
		}
		for ; i <= j && nums[i] == t; i++ {
		}
	}
	return res
}

func twoSum(nums []int, res *[][]int, target int) {
	if len(nums) < 2 {
		return
	}
	i, j := 0, len(nums)-1
	for ; i < j; {
		for ; i < j && nums[i]+nums[j] < target; i++ {
		}
		for ; i < j && nums[i]+nums[j] > target; j-- {
		}
		if i < j && nums[i]+nums[j] == target {
			*res = append(*res, []int{nums[i], nums[j]})
			t1, t2 := nums[i], nums[j]
			for ; i < j && t1 == nums[i]; i++ {
			}
			for ; i < j && t2 == nums[j]; j-- {
			}
		}
	}
}

File: leetcode/archived/C++/3Sum.cpp
class Solution {
	public:
		vector<vector<int> > threeSum(vector<int> &num) {
			vector<vector<int> > res;
			if(num.size() < 3)
				return res;

			sort(num.begin(), num.end());

			for(int i = 0; i < (num.size()-2); i++)
			{
				int l, h;
				int target = -num[i];
				l = i + 1;
				h = num.size()-1;

				while(l < h)
				{
					int sum = num[l] + num[h];
					if(sum < target)
						l++;
					else if(sum > target)
						h--;
					else
					{
						vector<int> triplet(3,0);
						triplet[0] = num[i];
						triplet[1] = num[l];
						triplet[2] = num[h];
						res.push_back(triplet);

						while((l < h) && (num[l] == triplet[1])) l++;

						while((l < h) && (num[h] == triplet[2])) h--;
					}
				}

				while ((i + 1 < num.size()) && (num[i + 1] == num[i])) 
					i++;

			}

			return res;
		}
};
